# <center> Rounding Numbers</center>
    
<center> (Reporting the correct number of significant figures.)</center>
<br><br><br><br>

#### <center> Don't Scroll Down Yet...</center>

---
**Question**: How would you round -2.45 to the first decimal place?

(Everyone type in their number in the "Chat box.")
<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

## Conceptual ideas

The act of rounding can introduce a bias into your numbers.

Consider 11 numbers that range from 0.5 -- 10.5 in 1.0 increments, round them to the ones place, and then take their mean.

In [1]:
import statistics

In [2]:
input_numbers = [number+0.5 for number in range(0, 11)]

print(f'Unrounded Numbers: {input_numbers}')
print(f'Average: {statistics.mean(input_numbers)}')

Unrounded Numbers: [0.5, 1.5, 2.5, 3.5, 4.5, 5.5, 6.5, 7.5, 8.5, 9.5, 10.5]
Average: 5.5


In [3]:
round_up = [number for number in range(1, 12)]

print(f'Rounded Up Numbers: {round_up}')
print(f'Average: {statistics.mean(round_up)}')

Rounded Up Numbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
Average: 6


In [4]:
round_down = [number for number in range(0, 11)]

print(f'Rounded Down Numbers: {round_down}')
print(f'Average: {statistics.mean(round_down)}')

Rounded Down Numbers: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Average: 5


What we observe is:
1. Rouding Up leads to 6, but
2. Rounding Down leads to 5.

**Therefore, the approach that we use to round numbers may result in a bias.**

---

## Conceptual approaches for rounding numbers

### 1. Bias methods

**Round Half Up** (i.e. bias towards **more positive**)
1.  7.6 $\rightarrow$ 8
2. **7.5 $\rightarrow$ 8**
3. 7.4 $\rightarrow$ 7
<br><br>
4. -7.6 $\rightarrow$ -8
5. **-7.5 $\rightarrow$ -7**
6. -7.4 $\rightarrow$ -7

**Round Half Down** (i.e. bias towards **more negative**)
1.  7.5 $\rightarrow$  7
2. -7.5 $\rightarrow$ -8

**Round Half To Zero** (i.e. bias towards **zero**)
1.  7.5 $\rightarrow$  7
2. -7.5 $\rightarrow$ -7

**Round Half Away from Zero** (i.e. bias towards **$\pm \infty$**)
1. 7.5 $\rightarrow$  8
2. -7.5 $\rightarrow$ -8

Additional information:

https://en.wikipedia.org/wiki/Rounding


### 2. Unbias methods (i.e. free from positive/negative bias and bias toward/away from zero)

**Round Half To Even** (i.e. zero is considered evenly distributed)
1.  7.5 $\rightarrow$  8
2.  8.5 $\rightarrow$  8
<br><br>
3. -7.5 $\rightarrow$ -8
4. -8.5 $\rightarrow$ -8


**Round Half To Odd** (i.e. zero is considered evenly distributed)
1.  7.5 $\rightarrow$  7
2.  8.5 $\rightarrow$  9
<br><br>
3. -7.5 $\rightarrow$ -7
4. -8.5 $\rightarrow$ -9

![rounding](00_images/rounding_bias.png)

---

## Different approaches within Python

---
### Python's built-in function 'round' (with examples)

https://docs.python.org/3/library/functions.html#round

Using the number from our figure above:

In [5]:
example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

for number in example_numbers:
    number_rounded = round(number)

    print(f'The number {number} becomes {number_rounded}.')

The number -4.5 becomes -4.
The number -3.5 becomes -4.
The number -0.5 becomes 0.
The number 0.5 becomes 0.
The number 3.5 becomes 4.
The number 4.5 becomes 4.


**Thus**, Python3's built-in ```round``` function uses **Round Half To Even**<br><br>


What happens if we push it slightly over 4.5?

In [6]:
print( round(4.51) )

5


So, the behavior is has it should be.

**Notice**: how ```round``` simpy returns the integer value of the float above.

Alternatively, one can specify the number of decimal places using ```round(number[, ndigits])```.

However, this does odd things to the number of digits that are returned (i.e. significant figures).

Let's round a list of floats to the second decimal place (i.e. to have 1 sigfig):

In [7]:
numbers_demo = [-0.055, 0.055, 0.05555, 10, 0.05555E-9]

for number in numbers_demo:
    number_rounded = round(number, 2)
    
    print(f'When rounded to 2 decimals, the number {number} becomes {number_rounded}.')

When rounded to 2 decimals, the number -0.055 becomes -0.06.
When rounded to 2 decimals, the number 0.055 becomes 0.06.
When rounded to 2 decimals, the number 0.05555 becomes 0.06.
When rounded to 2 decimals, the number 10 becomes 10.
When rounded to 2 decimals, the number 5.555e-11 becomes 0.0.


Notice two things:

1. 10, which as no decimals initially and thus a questionable number of sig figs, stays as 10 (good).
2. 0.05555E-9 has **4 sig figs**, but the function ```round``` returned **0.0**
    - should be **0.06E-9** (or 0.6E-10 or 6.E-11)
    - which means we **lost** some data information!

Okay, so ```round``` uses **Round Half To Even**

**However**, notice the following weird result when we round 2.675 to two decimal places.

In [8]:
print( round(2.675, 2) )

2.67


The ```round(2.675, 2)``` should result in **2.68**. Instead we obtained **2.67**!<br><br>

Another example of weirdness that helps us understand what is happening:

In [9]:
0.1 + 0.1 + 0.1

0.30000000000000004

In [10]:
0.1 + 0.1 + 0.1 - 0.3

5.551115123125783e-17

Reason: **floating-point representation error** (https://docs.python.org/3/tutorial/floatingpoint.html)

This is not a Python issue, but has to do with how computers store floating-point numbers in memory.

**Take-home message:**

From a perspective of being exact and correct, this is something that one needs to be aware of -- particularly when significant figures are important -- when developing code for scientific usage.

---
### Python's math library (with examples)

- https://docs.python.org/3/library/math.html

- math.ceil: **Round Up** (ceiling)
- math.floor: **Round Down**

In [11]:
import math

example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

for number in example_numbers:
    number_rounded = math.ceil(number)

    print(f'The number {number} becomes {number_rounded}.')

The number -4.5 becomes -4.
The number -3.5 becomes -3.
The number -0.5 becomes 0.
The number 0.5 becomes 1.
The number 3.5 becomes 4.
The number 4.5 becomes 5.


In [12]:
for number in example_numbers:
    number_rounded = math.floor(number)

    print(f'The number {number} becomes {number_rounded}.')

The number -4.5 becomes -5.
The number -3.5 becomes -4.
The number -0.5 becomes -1.
The number 0.5 becomes 0.
The number 3.5 becomes 3.
The number 4.5 becomes 4.


---
### The decimal library
- https://docs.python.org/3/library/decimal.html: "The decimal module provides support for fast correctly-rounded decimal floating point arithmetic."

Rounding options:
1. decimal.ROUND_CEILING $\rightarrow$ Rounding up
2. decimal.ROUND_FLOOR $\rightarrow$ Rounding down
3. decimal.ROUND_DOWN $\rightarrow$ Truncation
4. decimal.ROUND_UP $\rightarrow$ Rounding away from zero
5. decimal.ROUND_HALF_UP $\rightarrow$ Rounding half away from zero
6. decimal.ROUND_HALF_DOWN $\rightarrow$ Rounding half towards zero
7. decimal.ROUND_HALF_EVEN $\rightarrow$ Rounding half to even
8. decimal.ROUND_05UP $\rightarrow$ Rounding towards zero, unless the last digit is 0 or 5 then round away from zero

In [13]:
import decimal
from decimal import Decimal ## for improving readablility (otherwise, "decimal.Decimal")

Let's see what the **default settings** of ```decimal``` are:

In [14]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

Notice that the default rounding method is **ROUND_HALF_EVEN**.

Let's look at how Decimal handles **floats** and **strings**:

In [15]:
Decimal(3.14)

Decimal('3.140000000000000124344978758017532527446746826171875')

Side note: check the type that Decimal returns:

In [16]:
type(Decimal(3.14))

decimal.Decimal

In [17]:
Decimal('3.14')

Decimal('3.14')

Okay, now let's look at what happens if we add the number to itself.

We should get:

3.14 + 3.14 = 6.28

In [18]:
Decimal(3.14) + Decimal(3.14)

Decimal('6.280000000000000248689957516')

In [19]:
Decimal('3.14') + Decimal('3.14')

Decimal('6.28')

Therefore, using the string input approach for `Decimal` corrects the **floating-point representation error**.

---
**Side note**: we can revisit the case of `0.1 + 0.1 + 0.1` for fun and completeness:

In [20]:
0.1 + 0.1 + 0.1

0.30000000000000004

In [21]:
Decimal('0.1') + Decimal('0.1') + Decimal('0.1')

Decimal('0.3')

---
#### Rounding with decimals library

- done using ```quantized```:
    - ```Decimal(value="0", context=None).quantize(exp, rounding=None, context=None)```
    
    Decimal().quantize('0') --> ones place (e.g. 7)
    
    Decimal().quantize('0.0') --> ones and first decimal place (e.g. 7.1)
    
    Decimal().quantize('0.00') --> ones and first decimal place (e.g. 7.11)

In [22]:
example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

for number in example_numbers:
    number_rounded = Decimal(f'{number}').quantize(Decimal('0'))

    print(f'The number {number} becomes {number_rounded}.')

The number -4.5 becomes -4.
The number -3.5 becomes -4.
The number -0.5 becomes -0.
The number 0.5 becomes 0.
The number 3.5 becomes 4.
The number 4.5 becomes 4.


Thus, ```Decimal(value="0", context=None).quantize(exp, rounding=None, context=None)```:
- is **ROUND_HALF_EVEN**
- even thought `rounding=None`

Let's be more exact then and specify the rounding directly:

In [51]:
example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

for number in example_numbers:
    number_rounded = Decimal(f'{number}').quantize(Decimal('0'), rounding="ROUND_HALF_EVEN")

    print(f'The number {number} becomes {number_rounded}.')

The number -4.5 becomes -4.
The number -3.5 becomes -4.
The number -0.5 becomes -0.
The number 0.5 becomes 0.
The number 3.5 becomes 4.
The number 4.5 becomes 4.


We can now **change** the **number of decimal places**

In [48]:
example_demo = [-4.55, -4.65, 4.55, 4.65]

for number in example_demo:
    number_rounded = Decimal(f'{number}').quantize(Decimal('0.0'))

    print(f'The number {number} becomes {number_rounded}.')

The number -4.55 becomes -4.5.
The number -4.65 becomes -4.6.
The number 4.55 becomes 4.6.
The number 4.65 becomes 4.7.


In [49]:
example_demo = [-4.055, -4.065, 4.055, 4.065]

for number in example_demo:
    number_rounded = Decimal(f'{number}').quantize(Decimal('0.00'))

    print(f'The number {number} becomes {number_rounded}.')

The number -4.055 becomes -4.05.
The number -4.065 becomes -4.06.
The number 4.055 becomes 4.06.
The number 4.065 becomes 4.07.


Now let's change our rounding method to **rounding up** (i.e. ceiling).

In [26]:
decimal.getcontext().rounding = decimal.ROUND_CEILING
print(decimal.getcontext())

Context(prec=28, rounding=ROUND_CEILING, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[Inexact, FloatOperation, Rounded], traps=[InvalidOperation, DivisionByZero, Overflow])


In [50]:
example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

for number in example_numbers:
    number_rounded = Decimal(f'{number}').quantize(Decimal('1'))

    print(f'The number {number} becomes {number_rounded}.')

The number -4.5 becomes -4.
The number -3.5 becomes -3.
The number -0.5 becomes -0.
The number 0.5 becomes 1.
The number 3.5 becomes 4.
The number 4.5 becomes 5.


---
**Side note**: earlier we saw that ```round(2.675, 2)``` resulted in **2.67**, not the correct value of **2.68**. Let's see what `Decimal.quantize` does:

In [52]:
Decimal("2.675").quantize(Decimal("1.00"))

Decimal('2.68')

Therefore, we obtain the exact answer now (i.e. without the influence of floating point representation error).

---
Let now change the default **precision**. Precision here seem to be how many **digits** (i.e. not decimal places) are **available** for a number (default value is 28). **Note** that a number doesn't need to use as 28 digits, but they are there if needed.

In [29]:
Decimal('11') / Decimal('7')

Decimal('1.571428571428571428571428572')

Let's verify the precision by computing how characters the number has, which includes the decimal point:

In [30]:
len(str(Decimal('11') / Decimal('7')))

29

In [31]:
decimal.getcontext().prec=3
Decimal('11') / Decimal('7')

Decimal('1.58')

**Be careful**: This doesn't always alter the default settings of decimal. On my installed Jupyter-notebook (bug?), the default decimals revert back to 28. However, on Colaboratory the above command does alter the default settings.

In [32]:
Decimal('11') / Decimal('7')

Decimal('1.58')

In [33]:
decimal.getcontext()

Context(prec=3, rounding=ROUND_CEILING, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[Inexact, FloatOperation, Rounded], traps=[InvalidOperation, DivisionByZero, Overflow])

---
### The Numpy library (for consideration in the future when we do cover Numpy more)

In [34]:
import numpy

print( numpy.around(1.500) )
print()
print( numpy.around(1.500, decimals=1) )
print()
print( numpy.around(1.500, decimals=2) )

2.0

1.5

1.5


In [53]:
example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]

for number in example_numbers:
    number_rounded = numpy.around(number)

    print(f'The number {number} becomes {number_rounded}.')

The number -4.5 becomes -4.0.
The number -3.5 becomes -4.0.
The number -0.5 becomes -0.0.
The number 0.5 becomes 0.0.
The number 3.5 becomes 4.0.
The number 4.5 becomes 4.0.


In [54]:
example_demo = [-4.55, -4.65, -0.05, 0.05, 4.55, 4.65]

for number in example_demo:
    number_rounded = numpy.around(number, decimals=1)
    
    print(f'The number {number} becomes {number_rounded}.')

The number -4.55 becomes -4.6.
The number -4.65 becomes -4.6.
The number -0.05 becomes -0.0.
The number 0.05 becomes 0.0.
The number 4.55 becomes 4.6.
The number 4.65 becomes 4.6.


Thus, Numpy's `around` function is using the round to even method.

How about if we want to rounding up (ceiling) and down (floor) to the one's place? Numpy has that too.

Rounding Up:

In [55]:
for number in example_demo:
    number_rounded = numpy.ceil(number)
    
    print(f'The number {number} becomes {number_rounded}.')

The number -4.55 becomes -4.0.
The number -4.65 becomes -4.0.
The number -0.05 becomes -0.0.
The number 0.05 becomes 1.0.
The number 4.55 becomes 5.0.
The number 4.65 becomes 5.0.


Rounding Down:

In [56]:
for number in example_demo:
    number_rounded = numpy.floor(number)
    
    print(f'The number {number} becomes {number_rounded}.')

The number -4.55 becomes -5.0.
The number -4.65 becomes -5.0.
The number -0.05 becomes -1.0.
The number 0.05 becomes 0.0.
The number 4.55 becomes 4.0.
The number 4.65 becomes 4.0.


---
Now, there is one more way to consider...

### Formating print statements containing floats

https://docs.python.org/3.1/library/string.html#grammar-token-precision

In [80]:
## g: General format
##    - fixed point or exponential
print(f'{3.4500045:g}, {3.5500055:g}, {300000000005:g}')

3.45, 3.55001, 3e+11


In [72]:
## f: Fixed point
## rounding to odd number
##    demo: change 1f to 3f in on value
print(f'{3.45:0.1f}, {3.55:0.1f}, {-3.45:0.1f}, {-3.55:0.1f}, {300000000005:0.1f}')

3.5, 3.5, -3.5, -3.5, 300000000005.0


In [73]:
example_numbers = [-4.5, -3.5, -0.5, 0.5, 3.5, 4.5]


for number in example_numbers:
    print(f'The number {number} becomes {number:0.0f}.')

The number -4.5 becomes -4.
The number -3.5 becomes -4.
The number -0.5 becomes -0.
The number 0.5 becomes 0.
The number 3.5 becomes 4.
The number 4.5 becomes 4.


Seems to be **round to even**.

However, consider the following:

In [70]:
example_demo = [-4.55, -4.65, -0.05, 0.05, 4.55, 4.65]

for number in example_demo:
    print(f'The number {number} becomes {number:0.1f}.')

The number -4.55 becomes -4.5.
The number -4.65 becomes -4.7.
The number -0.05 becomes -0.1.
The number 0.05 becomes 0.1.
The number 4.55 becomes 4.5.
The number 4.65 becomes 4.7.


Now the behavour of formatting stings is **round to odd**.

In [69]:
## e: Exponent notation
## rounding to odd number
print(f'{3.45:0.1e}, {3.55:0.1e}, {300000000005:0.1e}')

3.5e+00, 3.5e+00, 3.0e+11


However, I'm not sure why 300000000005 -> 3.0e+11 and not 3.1e+11

In [74]:
example_demo = [-4.55, -4.65, -0.05, 0.05, 4.55, 4.65]

for number in example_demo:
    print(f'The number {number} becomes {number:0.1e}.')

The number -4.55 becomes -4.5e+00.
The number -4.65 becomes -4.7e+00.
The number -0.05 becomes -5.0e-02.
The number 0.05 becomes 5.0e-02.
The number 4.55 becomes 4.5e+00.
The number 4.65 becomes 4.7e+00.


Notice how **0.05** went to **5.0e-02** - i.e. a **gain of 1 sigfig**.

Even worse in the examples below:
1. 15 becomes 15.00 (i.e. a gain of sigfigs)
2. 1.555e-9 becomes 0.00 (ie. a loss of sigfigs)

In [75]:
number_list = [1.55, 15, 1.555E-9]

for number in number_list:
    print(f'When rounded to 2 decimals, the number {number} becomes {number:0.2f}.')

When rounded to 2 decimals, the number 1.55 becomes 1.55.
When rounded to 2 decimals, the number 15 becomes 15.00.
When rounded to 2 decimals, the number 1.555e-09 becomes 0.00.


In [77]:
for number in number_list:
    print(f'When rounded to 2 decimals, the number {number} becomes {number:0.2g}.')

When rounded to 2 decimals, the number 1.55 becomes 1.6.
When rounded to 2 decimals, the number 15 becomes 15.
When rounded to 2 decimals, the number 1.555e-09 becomes 1.6e-09.


So at least that is a little better.

## Take-home message

You have **think critically** about what you are doing and what your final goal is.

In other words, don't blindly take information that you find quickly and blindly use it.

**Create short test codes** to **verify** that **'what you think you are doing'** is actually **'what you are doing!'**

Consider using ```decimals``` or ```numpy.around``` (once we get to Numpy).

#### Sources:
1. https://docs.python.org/3/library/decimal.html#decimal.Decimal
2. https://realpython.com/python-rounding/#the-decimal-class